File: src/types/processor-interrupt.ts
import { NexLocation } from './api/locations';
import { DataConfirmation } from './data-confirmation';

export interface ProcessorInterrupt {
  type:
    | 'MISSING_API_KEY'
    | 'INVALID_API_KEY'
    | 'MISSING_SUBDOMAIN'
    | 'LOCATION_REQUIRED'
    | 'NO_LOCATIONS_FOUND'
    | 'MISSING_START_DATE'
    | 'MISSING_DAYS'
    | 'MISSING_APPOINTMENT_TYPE_NAME'
    | 'INTERNAL_ERROR'
    | 'NEEDS_CONFIRMATION'
    | 'PERMISSION_DENIED'
    | 'NOT_FOUND';
  resolutionData?: InterruptResolutionData;
}

type InterruptResolutionData =
  | { type: 'STRING'; payload: string }
  | { type: 'NUMBER'; payload: number }
  | { type: 'LOCATIONS'; payload: LocationResolutionData }
  | { type: 'CONFIRMATION'; payload: DataConfirmation }
  | { type: 'None'; payload: null };

type LocationResolutionData = {
  locations: NexLocation[];
  selected_location_ids?: number[];
};

File: src-tauri/src/services/processors/types/processor_interrupt.rs
use serde::Serialize;

use crate::{
    api::types::locations::Location,
    services::processors::types::data_confirmation::DataConfirmation,
};

#[derive(Serialize, Debug, Clone)]
#[serde(
    tag = "type",
    content = "resolutionData",
    rename_all = "SCREAMING_SNAKE_CASE"
)]
pub enum ProcessorInterrupt {
    MissingApiKey,
    InvalidApiKey,
    MissingSubdomain(Option<InterruptResolutionData>),
    LocationRequired(Option<InterruptResolutionData>),
    NoLocationsFound,
    MissingDays(Option<InterruptResolutionData>),
    MissingAppointmentTypeName(Option<InterruptResolutionData>),
    NeedsConfirmation(InterruptResolutionData),
    AcknowledgeCompletion(InterruptResolutionData),
    PermissionDenied(InterruptResolutionData),
    NotFound,
    InternalError(InterruptResolutionData),
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", content = "payload", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum InterruptResolutionData {
    String(String),
    Number(u32),
    Locations(LocationResolutionData),
    Confirmation(DataConfirmation),
    None,
}

#[derive(Debug, Clone, Serialize)]
pub struct LocationResolutionData {
    pub locations: Vec<Location>,
    pub selected_location_ids: Option<Vec<u32>>,
}
